### ISSUE TITLE
Emacs scrolls up by itself on mouse click or drag

### ISSUE THREAD

**Issue by foriequal0** [jwilm/alacritty#714]
Emacs scrolls up by itself when navigating cursor with mouse click or drag.
You might reproduce immediately on bare emacs (25 or 26)  by 
```sh
emacs -nw --no-init-file --no-loadup --eval '(xterm-mouse-mode t)' some-long-file
```
Vim doesn't with ```set mouse=a```, and Emacs with same flag on gnome-terminal, guake, xterm doesn't.

I built Alacritty using rustc 1.21.0-nightly (1d2a6df38 2017-08-03), Ubuntu 17.04.

**Comment by shicks** [comment]
This appears to be a problem with how alacritty sends SGR mouse codes.  After extensive debugging, I find that alacritty is inconsistent with xterm, gnome-terminal, and others.

Here is the test:
```
echo -e '\e[?1000h\e[?1006h'; showkey -a
```

Clicking and releasing the left button on the top-left of the terminal window in xterm produces (eliding the hex values)
```
^[[<0;1;1M
^[[<0;1;1m
```

But in alacritty the result is
```
^[[<0;1;1M
^[[<3;1;1m
```

Looking at input.rs, I see that the same `button` value is sent to both `normal_mouse_report` and to `sgr_mouse_report`.  But normal mouse reporting encodes the down/up state in the button value, while SGR uses the capital/lowercase `m` suffix for down/up.  So emacs is misinterpreting the mouseup as button 4, which it assumes is a mouse wheel (though from my experiments, xterm and alacritty both send button 64 and 65 for mouse wheels.  Alacritty doesn't send middle and right clicks at all, so it's hard to test those, but xterm sends 1, 2, and 3 for the other three buttons on my mouse, though, so it looks like it's just sending the index rather than a bitset).  Additionally, it seems that other terminals add various bitmasks for modifier keys.  It's difficult to get a consistent picture, because modified clicks on different terminals tend to bring up various menus, and I couldn't find any actual documentation anywhere.  But roughly, I gather the following:

* left: button = 0
* middle: button = 1
* right: button = 2
* any other button: button = 3
* wheel down: button = 64
* wheel up: button = 65
* shift: button |= 4
* alt: button |= 8
* ctrl: button |= 16
* motion: button |= 32

NOTE: I use "natural" scrolling via xinput, so my up/down scroll wheel might be backwards.

This is completely different from the lowest two bits of "normal mouse mode", which I gather is

* left: button = 0
* middle: button = 1
* right: button = 2
* release: button = 3

Alacritty misses the boat on a lot of the SGR codes.  Specifically

* as mentioned above, it incorrectly sends button=3 for release
* it does not include any modifiers, either in presses, releases, motion, or wheels
* it does not include the button in motion (though, again, since it doesn't send anything other than button 0, this could be argued as by design)

For now I'm patching my local version to just send button=0 instead of 3.  That fixes my immediate problem, but a good long-term solution.  I don't have the time at the moment to learn enough rust to do it correctly.

**Comment by Celti** [comment]
I've got an initial fix at Celti/alacritty — it reports the middle and right buttons properly, and reports releases properly in both SGR and normal mode. Since SGR mode condenses all other buttons down to a single code, and normal mode has no provision for them at all, I'm simply ignoring them at this time.

I don't pass modifiers in yet, nor handle buttons in motion yet.

**Comment by shicks** [comment]
Just because stock CLI apps don't expect Ctrl or Alt doesn't mean that there's no value in supporting them.  Emacs should be perfectly capable of binding to Ctrl+Mouse, at the very least.

**Comment by chrisduerr** [comment]
The #1141 PR fixes the SGR issues, adds modifiers and reports middle/right mouse button. Can anyone reproduce this issue with this PR or does it fix it?

Also I didn't see that @Celti already fixed this in #1057, I'll compare the two PRs and maybe we can put them together.
